### ISSUE TITLE
Phase 2: Complete Cross-Cell-Type Evaluation Matrix & Jaccard Overlap Analysis (M6.4)

### ISSUE THREAD

**Issue by purvanshjoshi** [PxA-Labs/interpretable-regulatory-genomics#19]
## Context

**Parent Issue**: #13 (Solidify Baselines & Gaps in Phases 1-3 to Align Codebase with Roadmap Specs)
**Related PR**: #17 (Phase 2 Deep Learning Multiclass & Generalization Upgrades)
**Related Discussion**: #18 (Phase 2 Solidification Results & Analysis)

---

## Background

As part of the Phase 2 solidification effort (PR #17), we successfully completed the **core deep learning upgrades** for multi-class classification and zero-shot cross-cell-type evaluation:

### What Has Been Completed ✅

| Deliverable | Status | Evidence |
| :--- | :--- | :--- |
| **3-class target architecture** (Background / PLS / dELS) | ✅ Merged | `src/models/cnn.py` — `ShallowCNN`, `DeepCNN`, `AttentionCNN` updated with `num_classes` parameter |
| **PyTorch training wrapper** for multiclass | ✅ Merged | `src/models/train_nn.py` — CrossEntropyLoss, softmax predictions, macro-averaged metrics |
| **GM12878 data curation** | ✅ Merged | `src/data/download.py` — cell-type specific annotation download (`ENCFF590IMH` for GM12878) |
| **K562 → GM12878 zero-shot evaluation** | ✅ Merged | Notebook achieves **58.55% zero-shot accuracy** (vs. 81.25% same-cell baseline) |
| **Unit tests** for multiclass & backwards compat | ✅ Merged | `tests/test_phase2_multiclass.py` |

### What Remains 🔲

This issue tracks the **remaining two sub-tasks** from Milestones **M6.4** (Experiment E4) and the Jaccard interpretation analysis that are required to fully close the cross-cell-type generalization experiment.

---

## Task 1: Complete the 2×2 Cross-Cell-Type Evaluation Matrix

**Milestone**: M6.4 (Cross-cell-type experiment — Experiment E4)
**Roadmap Reference**: [10-roadmap-milestones.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/10-roadmap-milestones.md)

### Current State

We currently have **2 of 4 cells** in the cross-cell-type evaluation matrix:

| | Eval: K562 | Eval: GM12878 |
| :--- | :--- | :--- |
| **Train: K562** | ✅ 81.25% (baseline) | ✅ 58.55% (zero-shot) |
| **Train: GM12878** | 🔲 Not started | 🔲 Not started |

### Required Work

1. **Train a model on GM12878 data**:
   - Use the GM12878 cCRE annotation file (`ENCFF590IMH`) already integrated in `src/data/download.py`
   - Extract GM12878-specific sequences using the existing `sequence_extractor.py` pipeline
   - Generate GC-matched negative samples using `negative_sampling.py`
   - Train `AttentionCNN` (best-performing architecture from K562) on GM12878 data using the same hyperparameters:
     - Learning rate: `1e-3`
     - Batch size: `64`
     - Epochs: `20` with early stopping (patience=5)
     - Chromosome 19 held-out test split
     - Random seed: `42`

2. **Evaluate the GM12878-trained model**:
   - **GM12878 → GM12878** (same-cell baseline): Record test accuracy, AUROC, AUPRC, macro-F1
   - **GM12878 → K562** (reverse zero-shot): Evaluate the GM12878-trained model on K562 test data

3. **Compile the full 2×2 matrix** with all four cells populated

### Expected Outputs

- [ ] Trained GM12878 model checkpoint saved under `experiments/logs/`
- [ ] Complete 2×2 accuracy matrix table
- [ ] Updated notebook or new notebook section documenting GM12878 training + evaluation
- [ ] `config.yaml` and `metrics.json` logged per [08-experiment-tracking-mlops.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/08-experiment-tracking-mlops.md) conventions
- [ ] Experiment registered as `exp_013_cnn_k562_gm12878_cross` in `experiments/REGISTRY.md`

---

## Task 2: Jaccard Overlap Analysis of Cell-Type-Specific vs. Agnostic Features

**Milestone**: M6.4 (Cross-cell-type experiment — Experiment E4, interpretation component)
**Roadmap Reference**: Section 2, Task 2.1 of Issue #13

### Description

Once both cell-type models (K562-trained and GM12878-trained) are available, extract the **top-20 most important features** from each model and perform a Jaccard similarity analysis to distinguish:

- **Cell-type-agnostic features**: Motifs/k-mers that appear in both models' top-20 lists (shared regulatory grammar)
- **Cell-type-specific features**: Motifs/k-mers unique to one cell type's top-20 list (cell-type-specific regulatory programs)

### Implementation Details

1. **Feature Extraction**:
   - For CNN models: Use first-layer convolutional filter weights → convert to PWMs → match to JASPAR motifs via Tomtom
   - For tree-based models (if also evaluated): Use SHAP TreeExplainer → extract top-20 features by mean |SHAP value|

2. **Jaccard Coefficient Calculation**:
   ```
   J(K562_top20, GM12878_top20) = |K562 ∩ GM12878| / |K562 ∪ GM12878|
   ```

3. **Biological Interpretation**:
   - Cross-reference shared motifs with known housekeeping TF binding sites (e.g., SP1, CTCF)
   - Cross-reference cell-type-specific motifs with known lineage-specific TFs:
     - K562 (erythroleukemia): GATA1, KLF1, TAL1
     - GM12878 (B-lymphoblastoid): PAX5, EBF1, PU.1

### Expected Outputs

- [ ] Top-20 feature list for K562-trained model
- [ ] Top-20 feature list for GM12878-trained model
- [ ] Jaccard similarity score and Venn diagram visualization
- [ ] Biological annotation table mapping shared/unique motifs to known TFs
- [ ] Summary paragraph suitable for inclusion in the research manuscript methods section

---

## Acceptance Criteria

- [ ] All 4 cells of the 2×2 cross-cell-type matrix are populated with accuracy, AUROC, and macro-F1 metrics
- [ ] Jaccard overlap analysis is complete with visualization
- [ ] Results are logged following MLOps conventions from [08-experiment-tracking-mlops.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/08-experiment-tracking-mlops.md)
- [ ] Notebook documentation is clear and reproducible on Kaggle T4 GPU environment
- [ ] Unit tests pass: `pytest tests/ -v`

---

## Labels

`phase-2`, `experiment`, `enhancement`

**Comment by purvanshjoshi** [comment]
### Related Discussions

This issue extends the work documented in the following discussions:

| Discussion | Description |
| :--- | :--- |
| **#4** — [Phase 2 Completed: Deep Learning Baselines, Convolutional Filters, and Attribution Maps](https://github.com/PxA-Labs/interpretable-regulatory-genomics/discussions/4) | Original Phase 2 completion discussion covering ShallowCNN/DeepCNN training, convolutional filter → JASPAR motif analysis, and Integrated Gradients attribution maps on K562 binary classification. |
| **#18** — [Phase 2 Solidification: Multi-class Deep Learning & Cross-Cell-Type Generalization (K562 → GM12878)](https://github.com/PxA-Labs/interpretable-regulatory-genomics/discussions/18) | Phase 2 solidification discussion documenting the 3-class upgrade (Background/PLS/dELS), GM12878 data curation, zero-shot evaluation (58.55% accuracy), and engineering insights from PR #17. |

The tasks in this issue (completing the 2×2 cross-cell-type matrix and Jaccard overlap analysis) directly build on the zero-shot evaluation infrastructure established in Discussion #18 and the CNN architectures validated in Discussion #4.

**Comment by purvanshjoshi** [comment]
## 📋 Full Documentation Audit — Phase 2 Remaining Gaps (Modeling & Experimentation)

After cross-referencing **all 6 project documentation files** against the actual codebase, the following additional gaps were identified beyond the original issue scope.

---

### 🔴 High Priority Gaps (Already Tracked Above)

| # | Gap | Source Document | Status |
|---|-----|-----------------|--------|
| 1 | Train GM12878 model + complete 2×2 matrix | M6.4 in [10-roadmap-milestones.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/10-roadmap-milestones.md), E4 in [04-research-design-document.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/04-research-design-document.md) | 🔲 Tracked in Task 1 above |
| 2 | Jaccard overlap analysis of cell-type features | E4 interpretation component, H3 hypothesis, Section 8.3 interpretability metrics in [04-research-design-document.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/04-research-design-document.md) | 🔲 Tracked in Task 2 above |

---

### 🟡 Medium Priority Gaps (New — From Documentation Audit)

#### Gap 3: 5-Class Element-Type Classification (vs. current 3-class)

**Source**: [05-dataset-strategy.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/05-dataset-strategy.md) Section 3.2

The documentation specifies a **5-class** label space for the element-type classification experiment (E5):

| Label | Definition | Current Status |
|-------|-----------|----------------|
| **PLS** (Promoter-Like Signature) | cCRE with high H3K4me3 and high DNase signal near TSS | ✅ Implemented as class 1 |
| **pELS** (Proximal Enhancer-Like) | cCRE with high H3K27ac, within 2 kb of TSS | 🔲 **Merged into dELS** |
| **dELS** (Distal Enhancer-Like) | cCRE with high H3K27ac, > 2 kb from TSS | ✅ Implemented as class 2 (combined with pELS) |
| **CTCF-only** | cCRE with CTCF binding but no other marks | 🔲 **Not implemented** |
| **Non-regulatory** | Sampled negative region | ✅ Implemented as class 0 |

**Decision needed**: Should we expand from 3-class to 5-class, or document the 3-class simplification as a deliberate design decision?

---

#### Gap 4: XGBoost Multiclass Evaluation for Experiment E5

**Source**: [06-modeling-roadmap.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/06-modeling-roadmap.md) Phase 2 Experiment Plan

The modeling roadmap specifies that element-type classification (E5) should be evaluated using **both** XGBoost and CNN architectures:

> *"Models: Evaluate using XGBoost and the ShallowCNN / DeepCNN architectures from src/models/cnn.py modified for multi-class cross-entropy loss."*

**Current state**: Only CNN architectures (ShallowCNN, DeepCNN, AttentionCNN) have been evaluated for multiclass. XGBoost multiclass evaluation has not been run.

**Required work**:
- Train XGBoost with multiclass objective (`multi:softmax` or `multi:softprob`) on the same 3-class data
- Compare XGBoost multiclass metrics (AUROC, macro-F1) against CNN multiclass metrics
- Run SHAP TreeExplainer on multiclass XGBoost to compare feature attributions with CNN saliency

---

#### Gap 5: Standalone Cross-Cell-Type Comparison Report

**Source**: [10-roadmap-milestones.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/10-roadmap-milestones.md) Phase 2 Deliverables List (item 3)

The roadmap lists "Cross-cell-type comparison report" as a standalone deliverable. Currently, cross-cell-type results exist only inside notebook outputs. A formal report summarizing the 2×2 matrix, motif overlap, and biological interpretation is required.

**Required work**:
- Create `docs/cross_cell_type_report.md` or equivalent
- Include the completed 2×2 accuracy matrix
- Include Jaccard overlap results and Venn diagram
- Include biological annotation of shared vs. cell-type-specific features

---

### 🟢 Low Priority Gaps (Code Organization)

#### Gap 6: `src/interpret/` Module Layer Missing

**Source**: [03-technical-design-document.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/03-technical-design-document.md) Section 1.2

The Technical Design Document specifies an **Interpretability Layer** with 4 modules:
- `src/interpret/importance.py` — Feature importance extraction
- `src/interpret/shap_analysis.py` — SHAP value computation
- `src/interpret/saliency.py` — Saliency map generation
- `src/interpret/motif_enrichment.py` — Motif cross-referencing

**Current state**: All interpretability code lives inline within Jupyter notebooks. No standalone modules exist.

#### Gap 7: `src/models/registry.py` Not Implemented

**Source**: [03-technical-design-document.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/03-technical-design-document.md) Section 3.1

The TDD specifies a `ModelRegistry` pattern with a `BaseModel` abstract class (`fit`, `predict`, `predict_proba`, `get_feature_importance`, `save`, `load`). This was never implemented — models are instantiated directly in notebooks.

#### Gap 8: Decision Gate 2→3 Formal Document

**Source**: [06-modeling-roadmap.md](https://github.com/PxA-Labs/interpretable-regulatory-genomics/blob/main/docs/06-modeling-roadmap.md) Decision Gate 2→3

All 5 gate criteria are **met** (CNN AUROC 0.8604 >= 0.80, >=3 filter-motif matches, saliency is biologically plausible, trade-off analysis documented, pretrained model loads on Kaggle). However, there is no **formal gate passage document** — the decision is implicit from Phase 3 execution.

---

### Summary: Updated Task List for Issue #19

| Priority | Task | Blocking? |
|----------|------|-----------|
| 🔴 | Complete 2×2 cross-cell-type matrix (GM12878 training) | Yes |
| 🔴 | Jaccard overlap analysis | Yes (blocked by above) |
| 🟡 | Decide on 5-class vs. 3-class label space | Design decision |
| 🟡 | XGBoost multiclass evaluation | No |
| 🟡 | Standalone cross-cell-type comparison report | No (blocked by matrix) |
| 🟢 | Extract interpretability code into `src/interpret/` modules | No |
| 🟢 | Implement `ModelRegistry` pattern | No |
| 🟢 | Write formal Decision Gate 2→3 document | No |

**Comment by purvanshjoshi** [comment]
## ⚠️ Important: Backward Compatibility Requirement

> **All contributors working on this issue must ensure 100% backward compatibility across the codebase.**

Any changes introduced as part of this issue — whether to model architectures, data pipelines, training wrappers, or utility modules — **must not break existing functionality**. The current binary classification pipeline, Phase 1 classical baselines, Phase 2 CNN training, and Phase 3 embedding workflows must all continue to function exactly as they do today.

### Guidelines

1. **Additive changes only**: New parameters (e.g., `num_classes`, `cell_type`) must have default values that preserve the existing binary/K562 behavior. Existing function signatures must not change in a way that breaks current callers.

2. **Run the full test suite before submitting a PR**:
   ```bash
   pytest tests/ -v
   ```
   All existing tests — including `test_phase2_multiclass.py` backward compatibility tests — must pass without modification.

3. **Verify notebook reproducibility**: Ensure that the existing notebooks (`phase1_classical_baseline.ipynb`, `phase2_deep_learning.ipynb`, `phase3_pretrained_embeddings.ipynb`) remain runnable and produce consistent results after your changes.

4. **No silent behavioral changes**: If a function's default behavior needs to change, introduce a new parameter with a default that preserves the legacy behavior, and update documentation accordingly.

5. **Test both code paths**: When adding multiclass or multi-cell-type support, explicitly test that the original binary classification and K562-only code paths still produce identical outputs.

This project has multiple interconnected phases, and regressions in earlier phases can cascade into downstream notebooks and experiments. Please treat backward compatibility as a hard requirement, not a suggestion.

**Comment by Saksham-dev-del** [comment]
Hi @purvanshjoshi,

I would like to work on this issue. My plan:

GM12878 training: Use the already-integrated ENCFF590IMH annotation in src/data/download.py, extract sequences via sequence_extractor.py, and generate GC-matched negatives via negative_sampling.py. Train AttentionCNN with the specified hyperparameters (lr=1e-3, batch_size=64, 20 epochs with early stopping at patience=5, chr19 holdout, seed=42).
Complete the 2×2 matrix: Record GM12878→GM12878 (same-cell) and GM12878→K562 (reverse zero-shot) results alongside the existing K562→K562 (81.25%) and K562→GM12878 (58.55%) figures, with AUROC, AUPRC, and macro-F1 for all four cells.
Jaccard overlap analysis: Extract top-20 filters from each cell-type model, convert to PWMs, and match to JASPAR via Tomtom. Compute J(K562_top20, GM12878_top20), cross-reference shared motifs against housekeeping TFs (SP1, CTCF) and cell-type-specific motifs against known lineage TFs (GATA1/KLF1/TAL1 for K562; PAX5/EBF1/PU.1 for GM12878), and produce a Venn diagram plus summary paragraph suitable for the manuscript methods section.
Log the run as exp_013_cnn_k562_gm12878_cross per MLOps conventions, and confirm pytest tests/ -v passes.

Could you please assign this issue to me? I'll open a PR once the matrix and analysis are complete.

**Comment by purvanshjoshi** [comment]
Hi @Saksham-dev-del,

Thanks for the detailed plan — you've clearly mapped out the full pipeline from data extraction to Jaccard analysis.

This is a **massive issue** with multiple interconnected tasks (GM12878 training, 2x2 matrix, Jaccard overlap, Venn diagram, MLOps logging). To keep the PR focused and reviewable, could you tell us which **specific part** you'd like to start with? For example:

- Just the GM12878 data extraction and model training?
- Just the 2x2 matrix compilation (if you want to skip training and work with existing outputs)?
- Just the Jaccard overlap analysis and Venn diagram?

Also, a few quick questions:
1. Do you have access to a GPU environment (Kaggle/Colab) for training the GM12878 model? Training CNNs on CPU will be very slow.
2. Are you comfortable with the backward compatibility requirement? All existing K562 pipelines must remain unbroken.

Let us know which piece you'd like to own and we'll get you assigned.
